File: benchmark/qwen3_asr/speed_bench.py
"""Qwen3-ASR speed benchmark (Python / moona3k).

Measures inference speed excluding model load time. Runs multiple iterations
for reliable timing data.

Usage:
    uv run --no-project \
        --with "mlx-qwen3-asr @ git+https://github.com/moona3k/mlx-qwen3-asr.git" \
        --with numpy \
        python vendors/dart-mlx-ffi/benchmark/qwen3_asr/speed_bench.py \
            --audio /tmp/speech_test_16k.wav [--iters 3]
"""

from __future__ import annotations

import argparse
import json
import struct
import sys
import time

import mlx.core as mx
import mlx.nn as nn
import numpy as np


def load_model_selective_quant(model_path: str):
    """Load moona3k model with selective quantization and fixed lm_head."""
    from mlx_qwen3_asr.config import Qwen3ASRConfig
    from mlx_qwen3_asr.load_models import (
        _load_safetensors,
        _resolve_path,
        remap_weights,
    )
    from mlx_qwen3_asr.model import Qwen3ASRModel

    resolved = _resolve_path(model_path)
    with open(resolved / "config.json") as f:
        raw_config = json.load(f)
    config = Qwen3ASRConfig.from_dict(raw_config)
    quant_cfg = raw_config.get("quantization", {})
    bits = int(quant_cfg.get("bits", 8))
    group_size = int(quant_cfg.get("group_size", 64))

    weights = _load_safetensors(resolved)
    weights = remap_weights(weights)
    model = Qwen3ASRModel(config)
    nn.quantize(model.model, bits=bits, group_size=group_size)
    model.load_weights(list(weights.items()), strict=False)

    if config.text_config.tie_word_embeddings:
        embed = model.model.embed_tokens
        dequant_w = mx.dequantize(
            embed.weight,
            embed.scales,
            embed.biases,
            group_size=group_size,
            bits=bits,
        )
        model.lm_head = nn.Linear(
            config.text_config.hidden_size,
            config.text_config.vocab_size,
            bias=False,
        )
        model.lm_head.weight = dequant_w

    mx.eval(model.parameters())
    model.eval()
    return model, config, str(resolved)


def parse_wav(path: str) -> np.ndarray:
    """Parse 16-bit PCM WAV to float32."""
    with open(path, "rb") as f:
        data = f.read()
    pos = 12
    while pos + 8 <= len(data):
        cid = data[pos : pos + 4]
        csz = struct.unpack_from("<I", data, pos + 4)[0]
        if cid == b"data":
            raw = data[pos + 8 : pos + 8 + csz]
            return np.frombuffer(raw, dtype=np.int16).astype(np.float32) / 32768.0
        pos += 8 + csz + (csz % 2)
    raise ValueError("No data chunk")


def bench_one(model, config, resolved_path: str, audio_np: np.ndarray) -> dict:
    """Run one full inference and return timing breakdown."""
    from mlx_qwen3_asr.audio import log_mel_spectrogram
    from mlx_qwen3_asr.generate import GenerationConfig, generate
    from mlx_qwen3_asr.tokenizer import Tokenizer

    tokenizer = Tokenizer(resolved_path)

    # Mel
    t0 = time.perf_counter()
    audio_mx = mx.array(audio_np)
    mel = log_mel_spectrogram(audio_mx)
    mx.eval(mel)
    mel_ms = (time.perf_counter() - t0) * 1000

    # Encoder
    t0 = time.perf_counter()
    n_frames = mel.shape[1]
    mel_b = mel[None, :, :]
    feat_lens = mx.array([n_frames])
    enc_out, _ = model.audio_tower(mel_b, feat_lens)
    mx.eval(enc_out)
    enc_ms = (time.perf_counter() - t0) * 1000

    # Prompt
    n_audio = enc_out.shape[1]
    prompt = tokenizer.build_prompt_tokens(
        n_audio_tokens=n_audio,
        language=None,
        context="",
    )
    input_ids = mx.array([prompt])
    seq_len = input_ids.shape[1]
    positions = mx.arange(seq_len)[None, :]
    position_ids = mx.stack([positions, positions, positions], axis=1)

    # Prefill + decode
    t0 = time.perf_counter()
    cache = model.create_cache(max_seq_len=seq_len + 448)
    prefill_logits = model.prefill(
        input_ids=input_ids,
        audio_features=enc_out,
        position_ids=position_ids,
        cache=cache,
    )
    mx.eval(prefill_logits)
    prefill_ms = (time.perf_counter() - t0) * 1000

    t0 = time.perf_counter()
    gen_config = GenerationConfig(max_new_tokens=448, temperature=0.0)
    gen_tokens = generate(
        model=model,
        input_ids=input_ids,
        audio_features=enc_out,
        position_ids=position_ids,
        config=gen_config,
    )
    mx.eval(mx.array(gen_tokens))
    total_gen_ms = (time.perf_counter() - t0) * 1000

    text = tokenizer.decode(gen_tokens)
    n_tokens = len(gen_tokens)

    return {
        "mel_ms": mel_ms,
        "encoder_ms": enc_ms,
        "prefill_ms": prefill_ms,
        "generate_ms": total_gen_ms,
        "total_inference_ms": mel_ms + enc_ms + total_gen_ms,
        "n_tokens": n_tokens,
        "tokens_per_sec": n_tokens / (total_gen_ms / 1000) if total_gen_ms > 0 else 0,
        "text_preview": text[:80],
    }


def main():
    parser = argparse.ArgumentParser(description="Qwen3-ASR Python speed benchmark")
    parser.add_argument("--audio", required=True, help="16kHz mono WAV path")
    parser.add_argument("--model", default="mlx-community/Qwen3-ASR-0.6B-8bit")
    parser.add_argument("--iters", type=int, default=3, help="Number of iterations")
    args = parser.parse_args()

    audio_np = parse_wav(args.audio)
    duration_s = len(audio_np) / 16000
    print(f"Audio: {len(audio_np)} samples ({duration_s:.2f}s)", file=sys.stderr)

    print("Loading model...", file=sys.stderr)
    t0 = time.perf_counter()
    model, config, resolved = load_model_selective_quant(args.model)
    load_ms = (time.perf_counter() - t0) * 1000
    print(f"Model loaded in {load_ms:.0f}ms", file=sys.stderr)

    # Warmup run
    print("Warmup run...", file=sys.stderr)
    _ = bench_one(model, config, resolved, audio_np)

    # Timed runs
    results = []
    for i in range(args.iters):
        print(f"Iteration {i + 1}/{args.iters}...", file=sys.stderr)
        r = bench_one(model, config, resolved, audio_np)
        results.append(r)
        print(
            f"  total={r['total_inference_ms']:.1f}ms  "
            f"mel={r['mel_ms']:.1f}ms  enc={r['encoder_ms']:.1f}ms  "
            f"gen={r['generate_ms']:.1f}ms  "
            f"tokens={r['n_tokens']}  "
            f"tok/s={r['tokens_per_sec']:.1f}",
            file=sys.stderr,
        )

    # Summary
    avg = {
        k: sum(r[k] for r in results) / len(results)
        for k in [
            "mel_ms",
            "encoder_ms",
            "prefill_ms",
            "generate_ms",
            "total_inference_ms",
            "tokens_per_sec",
        ]
    }
    avg["n_tokens"] = results[0]["n_tokens"]
    avg["audio_duration_s"] = duration_s
    avg["rtf"] = (avg["total_inference_ms"] / 1000) / duration_s
    avg["iters"] = args.iters
    avg["model_load_ms"] = load_ms
    avg["runtime"] = "python"
    avg["text_preview"] = results[0]["text_preview"]

    print(f"\n=== Python Benchmark Summary ({args.iters} iters) ===", file=sys.stderr)
    print(f"  Audio duration:    {duration_s:.2f}s", file=sys.stderr)
    print(f"  Model load:        {load_ms:.0f}ms", file=sys.stderr)
    print(f"  Mel:               {avg['mel_ms']:.1f}ms", file=sys.stderr)
    print(f"  Encoder:           {avg['encoder_ms']:.1f}ms", file=sys.stderr)
    print(f"  Prefill:           {avg['prefill_ms']:.1f}ms", file=sys.stderr)
    print(f"  Generate:          {avg['generate_ms']:.1f}ms", file=sys.stderr)
    print(f"  Total inference:   {avg['total_inference_ms']:.1f}ms", file=sys.stderr)
    print(f"  RTF:               {avg['rtf']:.4f}", file=sys.stderr)
    print(f"  Tokens:            {avg['n_tokens']}", file=sys.stderr)
    print(f"  Tokens/sec:        {avg['tokens_per_sec']:.1f}", file=sys.stderr)

    # JSON output to stdout
    print(json.dumps(avg, indent=2))


if __name__ == "__main__":
    main()

File: benchmark/qwen3_asr/speed_bench.dart
/// Qwen3-ASR speed benchmark (Dart / dart-mlx-ffi).
///
/// Measures inference speed excluding model load time, with multiple iterations
/// for reliable timing data. Outputs JSON to stdout, diagnostics to stderr.
///
/// Usage (from dart-mlx-ffi directory):
///   dart run benchmark/qwen3_asr/speed_bench.dart /tmp/speech_test_16k.wav [iters]
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:dart_mlx_ffi/dart_mlx_ffi.dart';
import 'package:dart_mlx_ffi/models.dart';

void main(List<String> args) {
  if (args.isEmpty) {
    stderr.writeln(
      'Usage: dart run benchmark/qwen3_asr/speed_bench.dart <wav> [iters]',
    );
    exitCode = 1;
    return;
  }

  final wavPath = args[0];
  final iters = args.length > 1 ? int.tryParse(args[1]) ?? 3 : 3;

  final audio = _parseWav(File(wavPath).readAsBytesSync());
  final durationS = audio.length / 16000;
  stderr.writeln(
    'Audio: ${audio.length} samples (${durationS.toStringAsFixed(2)}s)',
  );

  final home = Platform.environment['HOME'] ?? '';
  final modelPath =
      Platform.environment['CMDSPACE_QWEN3_ASR_MODEL']?.trim() ??
      '$home/.cmdspace/models/qwen3-asr/default';

  stderr.writeln('Loading model from $modelPath');
  final loadSw = Stopwatch()..start();
  final runner = Qwen3AsrRunner.load(modelPath);
  loadSw.stop();
  final loadMs = loadSw.elapsedMilliseconds;
  stderr.writeln('Model loaded in ${loadMs}ms');

  // Warmup run
  stderr.writeln('Warmup run...');
  runner.transcribe(audio, maxNewTokens: 448);

  // Get token count from a single run (to avoid double inference).
  final refIds = runner.transcribeToIds(audio, maxNewTokens: 448);
  final nTokens = refIds.length;

  // Timed runs — use transcribe() only (single inference per iteration).
  final results = <Map<String, dynamic>>[];
  for (var i = 0; i < iters; i++) {
    stderr.writeln('Iteration ${i + 1}/$iters...');
    final sw = Stopwatch()..start();
    final text = runner.transcribe(audio, maxNewTokens: 448);
    sw.stop();
    final totalMs = sw.elapsedMilliseconds;

    final tokPerSec = nTokens / (totalMs / 1000);
    results.add({
      'total_inference_ms': totalMs,
      'n_tokens': nTokens,
      'tokens_per_sec': tokPerSec,
      'text_preview': text.length > 80 ? text.substring(0, 80) : text,
    });
    stderr.writeln(
      '  total=${totalMs}ms  tokens=$nTokens  tok/s=${tokPerSec.toStringAsFixed(1)}',
    );
  }

  // Summary
  final avgTotalMs =
      results.fold<int>(0, (s, r) => s + (r['total_inference_ms'] as int)) /
      iters;
  final avgTokPerSec =
      results.fold<double>(0, (s, r) => s + (r['tokens_per_sec'] as double)) /
      iters;
  final rtf = (avgTotalMs / 1000) / durationS;

  stderr.writeln('\n=== Dart Benchmark Summary ($iters iters) ===');
  stderr.writeln('  Audio duration:    ${durationS.toStringAsFixed(2)}s');
  stderr.writeln('  Model load:        ${loadMs}ms');
  stderr.writeln('  Total inference:   ${avgTotalMs.toStringAsFixed(1)}ms');
  stderr.writeln('  RTF:               ${rtf.toStringAsFixed(4)}');
  stderr.writeln('  Tokens:            $nTokens');
  stderr.writeln('  Tokens/sec:        ${avgTokPerSec.toStringAsFixed(1)}');

  final summary = {
    'total_inference_ms': avgTotalMs,
    'n_tokens': nTokens,
    'tokens_per_sec': avgTokPerSec,
    'audio_duration_s': durationS,
    'rtf': rtf,
    'iters': iters,
    'model_load_ms': loadMs,
    'runtime': 'dart',
    'text_preview': results[0]['text_preview'],
  };
  stdout.writeln(jsonEncode(summary));

  runner.close();
}

Float32List _parseWav(Uint8List data) {
  if (data.length < 44) throw FormatException('WAV too short');
  final bd = ByteData.sublistView(data);
  var pos = 12;
  while (pos + 8 <= data.length) {
    final chunkId = String.fromCharCodes(data.sublist(pos, pos + 4));
    final chunkSize = bd.getUint32(pos + 4, Endian.little);
    if (chunkId == 'data') {
      final audioFormat = bd.getUint16(20, Endian.little);
      final bitsPerSample = bd.getUint16(34, Endian.little);
      final raw = data.sublist(pos + 8, pos + 8 + chunkSize);
      if (audioFormat == 1 && bitsPerSample == 16) {
        final samples = Float32List(raw.length ~/ 2);
        final rbd = ByteData.sublistView(raw);
        for (var i = 0; i < samples.length; i++) {
          samples[i] = rbd.getInt16(i * 2, Endian.little) / 32768.0;
        }
        return samples;
      } else if (audioFormat == 3 && bitsPerSample == 32) {
        return Float32List.sublistView(ByteData.sublistView(raw));
      }
      throw FormatException(
        'Unsupported: fmt=$audioFormat bits=$bitsPerSample',
      );
    }
    pos += 8 + chunkSize + (chunkSize % 2);
  }
  throw FormatException('No data chunk found');
}
